File: backend/services/scheduleCompiler.js
const MAX_TIMERS = 16;
const MAX_RULE_LENGTH = 511;

const ALL_DAYS = [0, 1, 2, 3, 4, 5, 6];

function toMinutes(hhmm) {
  if (typeof hhmm !== 'string') return null;
  const m = /^([01]\d|2[0-3]):([0-5]\d)$/.exec(hhmm);
  if (!m) return null;
  return parseInt(m[1], 10) * 60 + parseInt(m[2], 10);
}

function toHhmm(minutes) {
  const pad = (n) => String(n).padStart(2, '0');
  return `${pad(Math.floor(minutes / 60))}:${pad(minutes % 60)}`;
}

function tasmotaDayPosition(steesDay) {
  return (steesDay + 1) % 7;
}

function daysMask(steesDays) {
  const chars = ['0', '0', '0', '0', '0', '0', '0'];
  for (const day of steesDays) {
    chars[tasmotaDayPosition(day)] = '1';
  }
  return chars.join('');
}

function expandDays(recurrence) {
  if (!recurrence || recurrence.type === 'daily') return new Set(ALL_DAYS);
  if (recurrence.type !== 'custom') return null;
  const days = new Set();
  for (const d of recurrence.daysOfWeek || []) {
    if (Number.isInteger(d) && d >= 0 && d <= 6) days.add(d);
  }
  return days;
}

function scheduleLabel(schedule) {
  return String(schedule._id || schedule.name || '?');
}

function channelsForSchedule(schedule, deviceMax, conflicts, unsupportedReasons) {
  const channels = new Set();
  for (const c of schedule.channels || []) {
    if (Number.isInteger(c) && c >= 1 && c <= deviceMax) {
      channels.add(c);
    } else {
      conflicts.push(
        `Schedule ${scheduleLabel(schedule)} targets channel ${c} which is outside the device (1..${deviceMax})`,
      );
    }
  }
  if (channels.size === 0) {
    unsupportedReasons.push(
      `Schedule ${scheduleLabel(schedule)} has no valid channels after validation`,
    );
  }
  return channels;
}

function statesForDay(day, intervals) {
  const times = new Set();
  for (const iv of intervals) {
    times.add(iv.start);
    times.add(iv.end);
  }
  const sorted = Array.from(times).sort((a, b) => a - b);
  const coverageAt = (t) => {
    const on = new Set();
    for (const iv of intervals) {
      if (iv.start <= t && t < iv.end) {
        for (const c of iv.channels) on.add(c);
      }
    }
    return on;
  };
  const transitions = [];
  let previous = new Set();
  for (const t of sorted) {
    const next = coverageAt(t);
    const on = [];
    const off = [];
    for (const c of next) if (!previous.has(c)) on.push(c);
    for (const c of previous) if (!next.has(c)) off.push(c);
    on.sort((a, b) => a - b);
    off.sort((a, b) => a - b);
    if (on.length || off.length) transitions.push({ time: t, on, off });
    previous = next;
  }
  return transitions;
}

function scheduleKey(schedule) {
  const recurrence = schedule.recurrence || {};
  return JSON.stringify([
    schedule._id ? String(schedule._id) : '',
    schedule.name ? String(schedule.name) : '',
    schedule.enabled === true ? 1 : 0,
    (schedule.channels || []).slice().sort((a, b) => a - b),
    recurrence.type || '',
    (recurrence.daysOfWeek || []).slice().sort((a, b) => a - b),
    (schedule.timeRanges || []).map((r) => [r && r.start, r && r.end]),
  ]);
}

function compile({ deviceId, schedules, device }) {
  const conflicts = [];
  const unsupportedReasons = [];
  const deviceMax = device && Number.isInteger(device.channels) ? device.channels : 4;

  const enabled = (schedules || [])
    .filter((s) => s && s.enabled === true)
    .filter((s) => !deviceId || !s.deviceId || String(s.deviceId) === String(deviceId))
    .sort((a, b) => (scheduleKey(a) < scheduleKey(b) ? -1 : scheduleKey(a) > scheduleKey(b) ? 1 : 0));

  const intervals = [];
  for (const schedule of enabled) {
    const days = expandDays(schedule.recurrence);
    if (!days || days.size === 0) {
      unsupportedReasons.push(
        `Schedule ${scheduleLabel(schedule)} has an unsupported or empty recurrence`,
      );
      continue;
    }
    const channels = channelsForSchedule(schedule, deviceMax, conflicts, unsupportedReasons);
    if (channels.size === 0) continue;
    for (const range of schedule.timeRanges || []) {
      const start = toMinutes(range && range.start);
      const end = toMinutes(range && range.end);
      if (start === null || end === null || end <= start) {
        unsupportedReasons.push(
          `Schedule ${scheduleLabel(schedule)} has an invalid time range ${range && range.start}-${range && range.end}`,
        );
        continue;
      }
      for (const day of days) {
        intervals.push({ day, start, end, channels, scheduleId: String(schedule._id || '') });
      }
    }
  }

  const eventsByKey = new Map();
  for (const day of ALL_DAYS) {
    const dayIntervals = intervals.filter((iv) => iv.day === day);
    if (!dayIntervals.length) continue;
    for (const tr of statesForDay(day, dayIntervals)) {
      const key = `${tr.time}|${tr.on.join(',')}|${tr.off.join(',')}`;
      let event = eventsByKey.get(key);
      if (!event) {
        event = { time: tr.time, on: tr.on, off: tr.off, days: new Set(), sources: new Set() };
        eventsByKey.set(key, event);
      }
      event.days.add(day);
      for (const iv of dayIntervals) {
        if ((iv.start <= tr.time && tr.time < iv.end) || iv.end === tr.time) {
          if (iv.scheduleId) event.sources.add(iv.scheduleId);
        }
      }
    }
  }

  const events = Array.from(eventsByKey.values())
    .map((e) => ({
      time: e.time,
      on: e.on.slice(),
      off: e.off.slice(),
      days: Array.from(e.days).sort((a, b) => a - b),
      sources: Array.from(e.sources).sort(),
    }))
    .sort((a, b) => a.time - b.time || String(a.on).localeCompare(String(b.on)) || String(a.off).localeCompare(String(b.off)));

  const timers = [];
  const ruleClauses = [];
  events.forEach((event, i) => {
    const index = i + 1;
    const mask = daysMask(event.days);
    const actionCount = event.on.length + event.off.length;
    const isDirect = actionCount === 1;
    let config;
    if (isDirect) {
      const channel = event.on.length ? event.on[0] : event.off[0];
      config = {
        Enable: 1,
        Mode: 0,
        Time: toHhmm(event.time),
        Window: 0,
        Days: mask,
        Repeat: 1,
        Action: event.on.length ? 1 : 0,
        Output: channel,
      };
    } else {
      config = {
        Enable: 1,
        Mode: 0,
        Time: toHhmm(event.time),
        Window: 0,
        Days: mask,
        Repeat: 1,
        Output: 1,
        Action: 3,
      };
    }
    timers.push({
      index,
      config,
      event: { on: event.on.slice(), off: event.off.slice() },
      sources: event.sources.slice(),
    });
    if (!isDirect) {
      const commands = [];
      for (const c of event.on) commands.push(`Power${c} ON`);
      for (const c of event.off) commands.push(`Power${c} OFF`);
      const body = commands.length === 1 ? commands[0] : `Backlog ${commands.join('; ')}`;
      ruleClauses.push(`ON Clock#Timer=${index} DO ${body} ENDON`);
    }
  });

  const rules = [];
  const ruleText = ruleClauses.join(' ');
  if (ruleText) {
    rules.push({ ruleIndex: 1, length: ruleText.length, text: ruleText });
    if (ruleText.length > MAX_RULE_LENGTH) {
      unsupportedReasons.push(`Rule1 requires ${ruleText.length} chars (${MAX_RULE_LENGTH} max)`);
    }
  }

  const requiredTimerCount = timers.length;
  if (requiredTimerCount > MAX_TIMERS) {
    unsupportedReasons.push(`Requires ${requiredTimerCount} timers (${MAX_TIMERS} max)`);
  }

  return { deviceId, timers, rules, requiredTimerCount, conflicts, unsupportedReasons };
}

module.exports = { compile, MAX_TIMERS, MAX_RULE_LENGTH, daysMask };

File: local_app/lib/schedules/schedule_compiler.dart
import 'dart:convert';
import 'schedule_model.dart';

const int kMaxTimers = 16;
const int kMaxRuleLength = 511;
const int kSteesRuleIndex = 2;
const List<int> kAllDays = [0, 1, 2, 3, 4, 5, 6];

int? toMinutes(String? hhmm) {
  if (hhmm == null) return null;
  final m = RegExp(r'^([01]\d|2[0-3]):([0-5]\d)$').firstMatch(hhmm);
  if (m == null) return null;
  return int.parse(m.group(1)!) * 60 + int.parse(m.group(2)!);
}

String toHhmm(int minutes) {
  final p = (int n) => n.toString().padLeft(2, '0');
  return '${p(minutes ~/ 60)}:${p(minutes % 60)}';
}

int tasmotaDayPosition(int steesDay) => (steesDay + 1) % 7;

String daysMask(List<int> steesDays) {
  final chars = List.filled(7, '0');
  for (final day in steesDays) {
    chars[tasmotaDayPosition(day)] = '1';
  }
  return chars.join();
}

Set<int>? expandDays(LocalSchedule schedule) {
  if (schedule.recurrenceType == 'daily') return Set.of(kAllDays);
  if (schedule.recurrenceType != 'custom') return null;
  final days = <int>{};
  for (final d in schedule.daysOfWeek) {
    if (d >= 0 && d <= 6) days.add(d);
  }
  return days;
}

class TimerConfig {
  const TimerConfig({
    required this.enable,
    required this.mode,
    required this.time,
    required this.window,
    required this.days,
    required this.repeat,
    required this.output,
    required this.action,
  });

  final int enable;
  final int mode;
  final String time;
  final int window;
  final String days;
  final int repeat;
  final int output;
  final int action;

  static const TimerConfig factoryDefault = TimerConfig(
    enable: 0,
    mode: 0,
    time: '00:00',
    window: 0,
    days: '0000000',
    repeat: 0,
    output: 1,
    action: 0,
  );

  bool get isDefault =>
      enable == 0 &&
      mode == 0 &&
      time == '00:00' &&
      window == 0 &&
      days == '0000000' &&
      repeat == 0 &&
      action == 0;

  String toPayload() => jsonEncode({
        'Enable': enable,
        'Mode': mode,
        'Time': time,
        'Window': window,
        'Days': days,
        'Repeat': repeat,
        'Output': output,
        'Action': action,
      });

  static TimerConfig? fromDevice(Map<String, dynamic>? raw) {
    if (raw == null) return null;
    int asInt(Object? v) => v is num ? v.toInt() : 0;
    String time(Object? v) =>
        v is String && RegExp(r'^([01]\d|2[0-3]):[0-5]\d$').hasMatch(v) ? v : '00:00';
    String days(Object? v) =>
        v is String && RegExp(r'^[01]{7}$').hasMatch(v) ? v : '0000000';
    return TimerConfig(
      enable: asInt(raw['Enable']) == 1 ? 1 : 0,
      mode: asInt(raw['Mode']) == 1 ? 1 : 0,
      time: time(raw['Time']),
      window: asInt(raw['Window']),
      days: days(raw['Days']),
      repeat: asInt(raw['Repeat']),
      output: asInt(raw['Output']),
      action: asInt(raw['Action']),
    );
  }

  @override
  bool operator ==(Object other) =>
      other is TimerConfig &&
      enable == other.enable &&
      mode == other.mode &&
      time == other.time &&
      window == other.window &&
      days == other.days &&
      repeat == other.repeat &&
      output == other.output &&
      action == other.action;

  @override
  int get hashCode => Object.hash(enable, mode, time, window, days, repeat, output, action);
}

class PlannedTimer {
  const PlannedTimer({required this.index, required this.config, required this.on, required this.off});
  final int index;
  final TimerConfig config;
  final List<int> on;
  final List<int> off;
}

class CompiledPlan {
  const CompiledPlan({
    required this.timers,
    required this.ruleText,
    required this.requiredTimerCount,
    required this.conflicts,
    required this.unsupportedReasons,
  });
  final List<PlannedTimer> timers;
  final String ruleText;
  final int requiredTimerCount;
  final List<String> conflicts;
  final List<String> unsupportedReasons;

  bool get isValid => unsupportedReasons.isEmpty;
  bool get wantsRule => ruleText.isNotEmpty;
}

class _Interval {
  _Interval(this.day, this.start, this.end, this.channels, this.scheduleId);
  final int day;
  final int start;
  final int end;
  final Set<int> channels;
  final String scheduleId;
}

class _Transition {
  _Transition(this.time, this.on, this.off);
  final int time;
  final List<int> on;
  final List<int> off;
}

List<_Transition> _statesForDay(int day, List<_Interval> intervals) {
  final times = <int>{};
  for (final iv in intervals) {
    times.add(iv.start);
    times.add(iv.end);
  }
  final sorted = times.toList()..sort();
  Set<int> coverageAt(int t) {
    final on = <int>{};
    for (final iv in intervals) {
      if (iv.start <= t && t < iv.end) on.addAll(iv.channels);
    }
    return on;
  }

  final transitions = <_Transition>[];
  var previous = <int>{};
  for (final t in sorted) {
    final next = coverageAt(t);
    final on = next.where((c) => !previous.contains(c)).toList()..sort();
    final off = previous.where((c) => !next.contains(c)).toList()..sort();
    if (on.isNotEmpty || off.isNotEmpty) {
      transitions.add(_Transition(t, on, off));
    }
    previous = next;
  }
  return transitions;
}

CompiledPlan compileSchedules(List<LocalSchedule> schedules, {int deviceMaxChannels = 4}) {
  final conflicts = <String>[];
  final unsupportedReasons = <String>[];

  final enabled = schedules.where((s) => s.enabled).toList()
    ..sort((a, b) {
      final ka = '${a.id}|${a.name}';
      final kb = '${b.id}|${b.name}';
      return ka.compareTo(kb);
    });

  final intervals = <_Interval>[];
  for (final schedule in enabled) {
    final label = schedule.name.isEmpty ? schedule.id : schedule.name;
    final days = expandDays(schedule);
    if (days == null || days.isEmpty) {
      unsupportedReasons.add('Schedule $label has an unsupported or empty recurrence');
      continue;
    }
    final channels = <int>{};
    for (final c in schedule.channels) {
      if (c >= 1 && c <= deviceMaxChannels) {
        channels.add(c);
      } else {
        conflicts.add('Schedule $label targets channel $c which is outside the device (1..$deviceMaxChannels)');
      }
    }
    if (channels.isEmpty) {
      unsupportedReasons.add('Schedule $label has no valid channels after validation');
      continue;
    }
    for (final range in schedule.timeRanges) {
      final start = toMinutes(range.start);
      final end = toMinutes(range.end);
      if (start == null || end == null || end <= start) {
        unsupportedReasons.add('Schedule $label has an invalid time range ${range.start}-${range.end}');
        continue;
      }
      for (final day in days) {
        intervals.add(_Interval(day, start, end, channels, schedule.id));
      }
    }
  }

  final eventsByKey = <String, _MergedEvent>{};
  for (final day in kAllDays) {
    final dayIntervals = intervals.where((iv) => iv.day == day).toList();
    if (dayIntervals.isEmpty) continue;
    for (final tr in _statesForDay(day, dayIntervals)) {
      final key = '${tr.time}|${tr.on.join(',')}|${tr.off.join(',')}';
      final event = eventsByKey.putIfAbsent(
        key,
        () => _MergedEvent(tr.time, tr.on, tr.off),
      );
      event.days.add(day);
    }
  }

  final events = eventsByKey.values.toList()
    ..sort((a, b) {
      final c = a.time.compareTo(b.time);
      if (c != 0) return c;
      final d = a.on.join(',').compareTo(b.on.join(','));
      return d != 0 ? d : a.off.join(',').compareTo(b.off.join(','));
    });

  final timers = <PlannedTimer>[];
  final ruleClauses = <String>[];
  for (var i = 0; i < events.length; i++) {
    final event = events[i];
    final index = i + 1;
    final mask = daysMask(event.days.toList()..sort());
    final direct = event.on.length + event.off.length == 1;
    late final TimerConfig config;
    if (direct) {
      final channel = event.on.isNotEmpty ? event.on.first : event.off.first;
      config = TimerConfig(
        enable: 1,
        mode: 0,
        time: toHhmm(event.time),
        window: 0,
        days: mask,
        repeat: 1,
        output: channel,
        action: event.on.isNotEmpty ? 1 : 0,
      );
    } else {
      config = TimerConfig(
        enable: 1,
        mode: 0,
        time: toHhmm(event.time),
        window: 0,
        days: mask,
        repeat: 1,
        output: 1,
        action: 3,
      );
      final commands = [
        for (final c in event.on) 'Power$c ON',
        for (final c in event.off) 'Power$c OFF',
      ];
      final body = commands.length == 1 ? commands.first : 'Backlog ${commands.join('; ')}';
      ruleClauses.add('ON Clock#Timer=$index DO $body ENDON');
    }
    timers.add(PlannedTimer(index: index, config: config, on: event.on, off: event.off));
  }

  final ruleText = ruleClauses.join(' ');
  if (ruleText.isNotEmpty && ruleText.length > kMaxRuleLength) {
    unsupportedReasons.add('Rule2 requires ${ruleText.length} chars ($kMaxRuleLength max)');
  }
  if (timers.length > kMaxTimers) {
    unsupportedReasons.add('Requires ${timers.length} timers ($kMaxTimers max)');
  }

  return CompiledPlan(
    timers: timers,
    ruleText: ruleText,
    requiredTimerCount: timers.length,
    conflicts: conflicts,
    unsupportedReasons: unsupportedReasons,
  );
}

class _MergedEvent {
  _MergedEvent(this.time, this.on, this.off);
  final int time;
  final List<int> on;
  final List<int> off;
  final Set<int> days = {};
}
